Decorate Request and Reply

delvedor opened the pull request:
We want `Reply` and `Request` to have the same features of `decorate`: *encapsulation* and *inheritance*. There is a significant difference between these two object and Fastify itself, since Fastify is already instantiated, while Request and Reply are not.

This means that if we add methods to Fastify with `Object.create` we will not have troubles but with RR *(Request/Reply, ed.)* the approach is different, because every change must be applied to the `prototype`.

First if all, to be "moved around" (hooks callbacks, user handler...) RR must be in only one place, `fastify._Reply` for example.
This means that even if we use `Object.create` we will change that property for every node of the graph, and bye bye to encapsulation.

@mcollina and me had a chat at Codemotion Rome, and we thought that we can save RR inside the route node object, in this way we solve the encapsulation problem.

An effectively we solved it, this pr demonstrates that. But as you can see `decorateReply` is *synchronous*, this means that if we register a reply decorator inside a register we cannot access it, because the routes are declared synchronously and out of the `avvio` lifecycle. A solution could be use the `after` callback, but we are not consistent with the plugin import done in #53.
```js
fastify
  .register(require('reply-feature'))
  .after(() => {
    fastify.get(...)
  })
```

In addition if we think about it, we have not a real encapsulation here, but just a "snapshot" of Reply at the time the route is created.

Conclusion.
I don't know if this one is the solution, probably not, but it is a starting point for further discussion. I hope I didn't miss anything.
Feedback/ideas are welcome.

Thanks.
cc @jsumners @davidmarkclements

mcollina commented:
I don't think you are implementing this correctly.

Check this example:

```js

function A () {}

const current = build(A)

// this will be called after the following
// setTimeout has been scheduled but before
// it is triggered
process.nextTick(() => {
  current.prototype.myMethod = function () {
    console.log('aaaa')
  }
})

if (new A().myMethod) {
  console.error('kaboom')
  process.exit(1)
}

// simulate some asynchronicity
// such as an http request
setTimeout(() => {
  new current().myMethod()
}, 100)

function build (prev) {
  function B () {}

  B.prototype = new prev()

  return B
}
```

The method is added asynchronously, but the effects are still visibile.

I think the main problems with your implementation is that you are a) creating a new class each time, and b) copying over the properties.

As with `decorate`, a plugin author would need `Symbol.for('skip-override')` to modify the current context, and not have this affected only the derived plugins.

delvedor commented:
I think I'm missing something here, I've updated the code with your suggestions.
I also added `R.call(this, req, res, handle)` to the build function otherwise it will throw at runtime.
At the moment we have not the encapsulation, as you can see form the failing test.

jsumners commented:
I don't know what benefit this imparts. I don't know about Express, but I know Hapi doesn't have anything equivalent. That's not to say innovation isn't possible, I just think they would have considered such a feature and made a decision not to implement it for some good reason.

For me, the request hooks are powerful enough to accomplish anything I can currently think of that I would use this PR for. I will say that it would be nice to have a concept like https://hapijs.com/api#route-prerequisites though.

delvedor commented:
@jsumners sorry, I didn't get your point. Your are talking about the benefits of the encapsulation or the Request/Reply decorators?

In the first case the benefit is that you get a direct acyclic graph for your routes, this means no cross dependencies. A must have feature if you want to split the server in multiple services later without pain.

Talking about why the decorators, you can add some features to the fastify core objects, a templating support for example.

> For me, the request hooks are powerful enough to accomplish anything I can currently think of that I would use this PR for. I will say that it would be nice to have a concept like https://hapijs.com/api#route-prerequisites though.

I think that the route prerequisite is already supported by using the [preHandler hook](https://github.com/fastify/fastify/blob/master/docs/Hooks.md).

jsumners commented:
I'm asking about the benefit of the whole PR. I don't know what sort of scenarios it enables that aren't already possible with server level plugins/decorations.

mcollina commented:
The system is designed to support the same use case is there in Hapi, e.g. https://www.npmjs.com/package/lout as a sample module. Also https://www.npmjs.com/package/vision leverage a similar mechanism that is there in Hapi.

delvedor commented:
Well done @mcollina! :)

I'll do the same for `Request`.
We must also figure out how avoid the use of `R.call(this, req, res, handle)` since it is slow.

delvedor commented:
@mcollina if everything is fine to you I think we can proceed to merge and release! :)

mcollina commented on fastify.js:
```diff
@@ -98,7 +104,7 @@ function build (options) {
   function _runMiddlewares (err, req, res) {
     if (err) {
-      const reply = new Reply(req, res, null)
+      const reply = new fastify._Reply(req, res, null)
```
I think we can just use `Reply` here.

mcollina commented on fastify.js:
```diff
@@ -123,7 +129,7 @@ function build (options) {
   function middlewareCallback (err) {
     if (err) {
-      const reply = new Reply(this.req, this.res, null)
+      const reply = new fastify._Reply(this.req, this.res, null)
```
and here

mcollina commented on fastify.js:
```diff
@@ -132,7 +138,7 @@ function build (options) {
   function routeCallback (err) {
     if (err) {
-      const reply = new Reply(this.req, this.res, null)
+      const reply = new fastify._Reply(this.req, this.res, null)
```
and here.

mcollina commented on test/decorator.test.js:
```diff
@@ -54,3 +56,289 @@ test('addServerMethod should throw if a declared dependency is not present', t =
+    }))
+
+    fastify.get('/yes', (req, reply) => {
+      t.notOk(reply.test)
```
this should be defined from `instance`, and turned into `t.ok()`.

delvedor replied:
👍  same for the `decorateRequest` test
